src/main/java/ch/njol/skript/expressions/ExprSpawn.java

```diff
@@ -38,75 +31,83 @@
+
 	static {
-		Skript.registerExpression(ExprSpawn.class, Location.class, ExpressionType.PROPERTY, "[the] spawn[s] [(point|location)[s]] [of %worlds%]", "%worlds%'[s] spawn[s] [(point|location)[s]]");
+		Skript.registerExpression(ExprSpawn.class, Location.class, ExpressionType.PROPERTY,
+			"[the] spawn[s] [(point|location)[s]] [of %worlds%]",
+			"%worlds%'[s] spawn[s] [(point|location)[s]]"
+		);
 	}
-	
-	@SuppressWarnings({"unchecked", "null"})
+
 	@Override
-	public boolean init(final Expression<?>[] exprs, final int matchedPattern, final Kleenean isDelayed, final ParseResult parseResult) {
+	@SuppressWarnings("unchecked")
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
 		setExpr((Expression<? extends World>) exprs[0]);
 		return true;
 	}
-	
+
 	@Override
-	protected Location[] get(final Event e, final World[] source) {
-		if (getTime() == -1 && e instanceof SpawnChangeEvent && !Delay.isDelayed(e)) {
+	protected Location[] get(Event e, World[] source) {
+		if (getTime() == -1 && e instanceof SpawnChangeEvent && !Delay.isDelayed(e))
 			return new Location[] {((SpawnChangeEvent) e).getPreviousLocation()};
-		}
-		return get(source, new Converter<World, Location>() {
-			@Override
-			@Nullable
-			public Location convert(final World w) {
-				return w.getSpawnLocation();
-			}
-		});
+		return get(source, World::getSpawnLocation);
 	}
-	
-	@Override
-	public Class<? extends Location> getReturnType() {
-		return Location.class;
-	}
-	
-	@Override
-	public String toString(final @Nullable Event e, final boolean debug) {
-		return "spawn of " + getExpr().toString(e, debug);
-	}
-	
+
 	@Override
 	@Nullable
-	public Class<?>[] acceptChange(final ChangeMode mode) {
+	public Class<?>[] acceptChange(ChangeMode mode) {
 		if (mode == ChangeMode.SET)
 			return CollectionUtils.array(Location.class);
 		return null;
 	}
-	
+
 	@Override
-	public void change(final Event e, final @Nullable Object[] delta, final ChangeMode mode) {
-		assert mode == ChangeMode.SET;
-		assert delta != null;
-		
-		final Location l = (Location) delta[0];
-		final int x = l.getBlockX(), y = l.getBlockY(), z = l.getBlockZ();
-		for (final World w : getExpr().getArray(e)) {
-			w.setSpawnLocation(x, y, z);
+	public void change(Event e, @Nullable Object[] delta, ChangeMode mode) {
+		//noinspection ConstantConditions
+		if (delta == null)
+			return;
+
+		Location originalLocation = (Location) delta[0];
+		assert originalLocation != null;
+		for (World world : getExpr().getArray(e)) {
+			Location location = originalLocation.clone();
+			if (location.getWorld() == null) {
+				location.setWorld(world);
+				world.setSpawnLocation(location);
+			} else if (location.getWorld().equals(world)) {
+				world.setSpawnLocation(location);
+			}
 		}
 	}
-	
-	@SuppressWarnings("unchecked")
+
 	@Override
-	public boolean setTime(final int time) {
+	@SuppressWarnings("unchecked")
```

Comment:
Why did this need unchecked suppress warning? Does setTime not have `@SafeArgs` VarArgs

Reply:
Correct. Probably something we should take a look at.